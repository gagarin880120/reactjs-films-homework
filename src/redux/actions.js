function searchResultsAction(results) {
  return {
      type: 'SEARCH',
      results: results
  };
}

function genresAction(genres) {
  return {
      type: 'GENRES',
      genres: genres
  };
}

export { searchResultsAction, genresAction };
