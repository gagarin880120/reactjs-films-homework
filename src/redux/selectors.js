import getGenresString from '../helpers/helpers';

const searchResultsSelector = (state) => {
  state.searchResults.map((v) => {
    const obj = v;
    obj.genres = getGenresString(v.genre_ids, state.genres);
    return { ...obj };
  });
  return state.searchResults;
};

const modalSelector = (state) => state.isModalOpen;

const trailerSelector = (state) => state.trailerURL;


export { searchResultsSelector, modalSelector, trailerSelector };
