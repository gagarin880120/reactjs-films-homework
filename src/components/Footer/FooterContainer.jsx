import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Footer from './Footer';
import { getMovies, getMoviesBySearch, getMoviesByGenre } from '../../redux/actions';
import {
  resultsSelector, currentPageSelector, currentAPIRequestSelector,
  areMoviesLoadedSelector, totalPagesSelector,
} from '../../redux/selectors';

export function FooterContainer({
  results, onIntersect, currentPage,
  areMoviesLoaded, totalPages, currentAPIRequest,
}) {
  return (
    <Footer
      results={results}
      currentPage={currentPage}
      onIntersect={onIntersect}
      areMoviesLoaded={areMoviesLoaded}
      totalPages={totalPages}
      currentAPIRequest={currentAPIRequest}
    />
  );
}

export const mapStateToProps = (state) => ({
  results: resultsSelector(state),
  currentPage: currentPageSelector(state),
  areMoviesLoaded: areMoviesLoadedSelector(state),
  totalPages: totalPagesSelector(state),
  currentAPIRequest: currentAPIRequestSelector(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onIntersect(currentAPIRequest, page, results) {
    if (currentAPIRequest.includes('search')) {
      dispatch(getMoviesBySearch(currentAPIRequest, page, results));
    } else if (currentAPIRequest.includes('genre')) {
      dispatch(getMoviesByGenre(currentAPIRequest, page, results));
    } else {
      dispatch(getMovies(currentAPIRequest, page, results));
    }
  },
});

FooterContainer.propTypes = {
  results: PropTypes.instanceOf(Array),
  currentPage: PropTypes.number,
  onIntersect: PropTypes.func,
  areMoviesLoaded: PropTypes.bool,
  totalPages: PropTypes.number,
  currentAPIRequest: PropTypes.string,
};

FooterContainer.defaultProps = {
  results: [],
  currentPage: 1,
  onIntersect: null,
  areMoviesLoaded: false,
  totalPages: 0,
  currentAPIRequest: '',
};

export default connect(mapStateToProps, mapDispatchToProps)(FooterContainer);
