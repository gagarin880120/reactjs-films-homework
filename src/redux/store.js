import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const initialState = {
  searchResults: [],
  genres: [],
  isModalOpen: false,
  trailerURL: '',
  isTrailerLoaded: false,
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case 'SEARCH':
      return { ...state, searchResults: action.results };
    case 'GENRES':
      return { ...state, genres: action.genres };
    case 'MODAL':
      return { ...state, isModalOpen: action.isModalOpen };
    case 'TRAILER':
      return { ...state, trailerURL: action.trailerURL };
    case 'IS_TRAILER_LOADED':
      return { ...state, isTrailerLoaded: action.isTrailerLoaded };
    default:
      return state;
  }
}

export const store = createStore(reducer, applyMiddleware(thunk));
