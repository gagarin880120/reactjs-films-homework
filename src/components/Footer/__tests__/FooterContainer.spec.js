import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import { FooterContainer, mapStateToProps, mapDispatchToProps } from '../FooterContainer';
import { getMovies } from '../../../redux/actions';

jest.mock('../../../redux/actions',() => ({
  getMovies: jest.fn().mockReturnValue('getMoviesAction'),
}))

describe('FooterContainer component', () => {
  test('should render without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<FooterContainer />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    test('should return the right value', () => {

      const initialState = {
        results: [{title: 'Matrix', genre_ids: [28]}],
        currentPage: 1,
        query: 'query',
        areMoviesLoaded: false,
        currentURL: 'url',
        totalPages: 10,
        genres: [{id: 28, name: 'Action'}],
      };

      expect(mapStateToProps(initialState).results).toEqual([{title: 'Matrix', genre_ids: [28], genres: 'Action'}]);
      expect(mapStateToProps(initialState).currentPage).toEqual(1);
      expect(mapStateToProps(initialState).query).toEqual('query');
      expect(mapStateToProps(initialState).areMoviesLoaded).toEqual(false);
      expect(mapStateToProps(initialState).url).toEqual('url');
      expect(mapStateToProps(initialState).totalPages).toEqual(10);
    });
  });

  describe('mapDispatchToProps', () => {
    afterEach(() => {
      jest.clearAllMocks();
    })

    test('onIntersect should dispatch getMovies action', () => {
      const dispatch = jest.fn();
      const {onIntersect} = mapDispatchToProps(dispatch);

      onIntersect();

      expect(getMovies).toHaveBeenCalled();
    });
  });
});
