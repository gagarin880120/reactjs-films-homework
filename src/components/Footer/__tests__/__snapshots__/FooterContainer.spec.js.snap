// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FooterContainer component should render without crashing 1`] = `
<Footer
  areMoviesLoaded={false}
  currentAction=""
  currentPage={1}
  onIntersect={null}
  query=""
  results={Array []}
  totalPages={0}
  url=""
/>
`;
